plugins/inputs/tencentcloudcm/tencentcloudcm.go

```diff
@@ -0,0 +1,371 @@
+						namespace.Name,
+					)
+				}
+			}
+		}
+	}
+
+	// start discovery
+	go t.discoverTool.Discover(t.Accounts, t.DiscoveryInterval, t.Endpoint)
+	t.discoverTool.DiscoverMetrics()
+	for i := range t.Accounts {
+		for j := range t.Accounts[i].Namespaces {
+			// use discovered metrics if not specified
+			if len(t.Accounts[i].Namespaces[j].Metrics) == 0 {
+				metrics, ok := t.discoverTool.DiscoveredMetrics[t.Accounts[i].Namespaces[j].Name]
+				if !ok {
+					return fmt.Errorf("unsupported namespace %s for discovering metrics, please specify metrics",
+						t.Accounts[i].Namespaces[j].Name)
+				}
+				t.Accounts[i].Namespaces[j].Metrics = metrics
+			}
+		}
+	}
+
+	t.client = &cloudmonitorClient{Accounts: t.Accounts}
+
+	return nil
+
+}
+
+func (t *TencentCloudCM) updateWindow(relativeTo time.Time) {
+
+	windowEnd := relativeTo.Add(-time.Duration(t.Delay))
+
+	// starting point is two times the aggregation period to make sure all points are covered
+	t.windowStart = windowEnd.Add(-time.Duration(t.Period) * 2)
+
+	t.windowEnd = windowEnd
+}
+
+func (t *TencentCloudCM) Gather(acc telegraf.Accumulator) error {
+	t.updateWindow(time.Now())
+
+	metricObjects := t.client.GetMetricObjects(*t)
+
+	lmtr := limiter.NewRateLimiter(t.RateLimit, time.Second)
+	defer lmtr.Stop()
+
+	wg := sync.WaitGroup{}
+	rLock := sync.Mutex{}
+	results := []monitor.GetMonitorDataResponse{}
+
+	// requestIDMap contains request ID and metric objects for later aggregation
+	requestIDMap := map[string]metricObject{}
+	for _, obj := range metricObjects {
+
+		wg.Add(1)
+		<-lmtr.C
+		go func(m metricObject) {
+			defer wg.Done()
+
+			for {
+
+				client, err := t.client.NewClient(m.Region, m.Account.crs, *t)
+				if err != nil {
+					acc.AddError(err)
+					break
+				}
+
+				size := 100
+				if len(m.MonitorInstances) >= size {
+					batch := m.MonitorInstances[:size]
+					if len(batch) == 0 {
+						break
+					}
+
+					request := t.client.NewGetMonitorDataRequest(m.Namespace, m.Metric, batch, *t)
+
+					result, err := t.client.GatherMetrics(client, request, *t)
+					if err != nil {
+						acc.AddError(err)
+						break
+					}
+
+					rLock.Lock()
+					requestIDMap[*result.Response.RequestId] = m
+					results = append(results, *result)
+					rLock.Unlock()
+
+					m.MonitorInstances = m.MonitorInstances[size:]
+				} else {
+					request := t.client.NewGetMonitorDataRequest(m.Namespace, m.Metric, m.MonitorInstances, *t)
+
+					result, err := t.client.GatherMetrics(client, request, *t)
+					if err != nil {
+						acc.AddError(err)
+						break
+					}
+
+					rLock.Lock()
+					requestIDMap[*result.Response.RequestId] = m
+					results = append(results, *result)
+					rLock.Unlock()
+					break
+				}
+
+			}
+
+		}(obj)
+
+	}
+	wg.Wait()
+	for _, result := range results {
+		for _, datapoints := range result.Response.DataPoints {
+			tags := map[string]string{}
+
+			keys := []string{}
```

Comment:
What do you need those `keys` for?

Reply:
`keys` are constructed using specific dimension value to uniquely identify and access the instance for detailed instance information which is later added to tags.

Reply:
Right, you are constructing this array, but I cannot see where you are doing something with it. As far as I can see, it is only used in line 325 (`keys = append(keys,...)` but not anywhere else. Am I missing something?

Reply:
It is used as param for function `newKey()` in line 331, as follow:

```go
instance := t.discoverTool.GetInstance(metricObject.Account.Name, metricObject.Namespace, metricObject.Region, newKey(strings.Join(keys, "-")))
```

Reply:
Oh I missed that one. Sorry!